core/server/models/post.js

```diff
@@ -676,24 +678,56 @@ Post = ghostBookshelf.Model.extend({
             return unsafeAttrs[attr] && unsafeAttrs[attr] !== postModel.get(attr);
         }
 
-        function actorIsAuthor(loadedPermissions) {
-            return loadedPermissions.user && _.some(loadedPermissions.user.roles, {name: 'Author'});
-        }
-
         function isOwner() {
             return unsafeAttrs.author_id && unsafeAttrs.author_id === context.user;
         }
 
-        if (actorIsAuthor(loadedPermissions) && action === 'edit' && isChanging('author_id')) {
-            hasUserPermission = false;
-        } else if (actorIsAuthor(loadedPermissions) && action === 'add') {
+        function isCurrentOwner() {
+            return context.user === postModel.get('author_id');
+        }
+
+        function isPublished() {
+            return unsafeAttrs.status && unsafeAttrs.status !== 'draft';
+        }
+
+        function isDraft() {
+            return postModel.get('status') === 'draft';
+        }
+
+        isContributor = baseUtils.actorIs(loadedPermissions.user, 'Contributor');
+        isAuthor = baseUtils.actorIs(loadedPermissions.user, 'Author');
+        isEdit = (action === 'edit');
+        isAdd = (action === 'add');
+        isDestroy = (action === 'destroy');
+
+        if (isContributor && isEdit) {
+            // Only allow contributor edit if neither status or author id are changing, and the post is a draft post
+            hasUserPermission = !isChanging('status') && !isChanging('author_id') && isDraft();
```

Comment:
We don't check that the contributor owns this post? 🤔 
Have not tested it.

Reply:
Oops 😱 good catch.